### DISCUSSION TITLE
What's the recommended way to iterate over `PyMapping` object?

### DISCUSSION THREAD

**Discussion by tushushu** [PyO3/pyo3#2718] [Questions]
I thought I could iterate over `PyMapping` object like this:
```rust
for (key, value) in mapping.iter()
```
But actually I cannot, and the below codes explain what is inside the `PyMapping` object:
```rust
fn foo(d: &PyMapping){
    println!("d.iter() is PyResult<&PyIterator> type.");
    if let Ok(it) = d.iter(){
        println!("x is &PyIterator type.");
        for x in it{
            println!("x is Result<&PyAny, PyErr> type.");
            if let Ok(y) = x{
                println!("y is &PyAny type.");
            }
        }
    }
}
```
So it seems that the `iter` method isn't something I am finding. Is there any suggestion?
Thanks very much!

**Comment by samuelcolvin** [comment]
I think you need `d.items().iter()`, see [here](https://github.com/pydantic/pydantic-core/blob/main/src/input/input_python.rs#L651)

**Reply by tushushu** [reply]
Thanks for the answer. It seems that the `items` method is eager executed, and sometimes the for loop may need break in the middle, which does not require iterating over all the elements in the `PyDict` object. There might be some extra performance overhead?

**Reply by tushushu** [reply]
By the way, I tried `items().iter()`, and still need to use multiple layers of `if let` to "extract" the element of the `PyMapping` object. So the further question could be:

- In what kind of case, the iterating over `PyMapping` should return an `PyErr`? Shouldn't this data structure be always iterable?

**Reply by davidhewitt** [reply]
The `PyMapping_Items` C API this uses is indeed eager, it returns a list: https://docs.python.org/3/c-api/mapping.html#c.PyMapping_Items

In general we can't offer anything concrete methods in `PyO3` which differ from this. However you could call the Python "items" method which might be support lazy iteration (e.g. for dict):

```rust
fn foo(d: &PyMapping) -> PyResult<()> {
    for item in d.call_method0("items")?.iter()? {
        let (key, value) = item.extract()?;
        // do your thing
    }
    Ok(())
}
```

**Reply by tushushu** [reply]
Thanks @davidhewitt , let me also try the way you suggested.
